pkg/controller/direct/logging/utils.go

```diff
@@ -0,0 +1,142 @@
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/clients/generated/apis/logging/v1beta1"
+	"github.com/googleapis/gax-go/v2/apierror"
+	api "google.golang.org/api/logging/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+)
+
+// todo acpana: add to factor out to top level package
+// todo acpana: begin
+func ValueOf[T any](p *T) T {
+	var v T
+	if p != nil {
+		v = *p
+	}
+	return v
+}
+
+// IsNotFound returns true if the given error is an HTTP 404.
+func IsNotFound(err error) bool {
+	return HasHTTPCode(err, 404)
+}
+
+// HasHTTPCode returns true if the given error is an HTTP response with the given code.
+func HasHTTPCode(err error, code int) bool {
+	if err == nil {
+		return false
+	}
+	apiError := &apierror.APIError{}
+	if errors.As(err, &apiError) {
+		if apiError.HTTPCode() == code {
+			return true
+		}
+	} else {
+		klog.Warningf("unexpected error type %T", err)
+	}
+	return false
+}
+
+func setStatus(u *unstructured.Unstructured, typedStatus any) error {
+	status, err := runtime.DefaultUnstructuredConverter.ToUnstructured(typedStatus)
+	if err != nil {
+		return fmt.Errorf("error converting status to unstructured: %w", err)
+	}
+	u.Object["status"] = status
+
+	return nil
+}
+
+// todo acpana: end common things
+
+// todo acpana: house this somewhere else
+func compareBucketOptions(kccObj *krm.LogmetricBucketOptions, apiObj *api.BucketOptions) bool {
+	if kccObj == nil || apiObj == nil {
+		return false
+	}
+
+	if kccObj.ExplicitBuckets == nil || apiObj.ExplicitBuckets == nil {
+		return false
+	}
+	if !reflect.DeepEqual(kccObj.ExplicitBuckets.Bounds, apiObj.ExplicitBuckets.Bounds) {
+		return false
+	}
+
+	if kccObj.ExponentialBuckets == nil || apiObj.ExponentialBuckets == nil {
+		return false
+	}
+	kccApiReducedExponentialBuckets := struct { // todo acpana: double check that reduction does not violate "equality"
+		growthFactor     float64
+		numFiniteBuckets int64
+		scale            float64
+	}{
+		growthFactor:     apiObj.ExponentialBuckets.GrowthFactor,
+		numFiniteBuckets: apiObj.LinearBuckets.NumFiniteBuckets,
+		scale:            apiObj.ExponentialBuckets.Scale,
+	}
+	if !reflect.DeepEqual(kccObj.ExponentialBuckets, kccApiReducedExponentialBuckets) {
+		return false
+	}
+
+	if kccObj.LinearBuckets == nil || apiObj.LinearBuckets == nil {
+		return false
+	}
+	kccApiReducedLinearBuckets := struct {
+		numFiniteBuckets int64
+		offset           float64
+		witdh            float64
+	}{
+		numFiniteBuckets: apiObj.LinearBuckets.NumFiniteBuckets,
+		offset:           apiObj.LinearBuckets.Offset,
+		witdh:            apiObj.LinearBuckets.Width,
+	}
+
+	return reflect.DeepEqual(kccObj.LinearBuckets, kccApiReducedLinearBuckets)
+}
+
+func expandKCCtoAPIForBucketOptions(kccObj *krm.LogmetricBucketOptions) *api.BucketOptions {
+	if kccObj == nil {
+		return nil
+	}
+
+	expanded := &api.BucketOptions{}
```

Comment:
I like this, I'm starting to imagine the auto-generated version of this.

```
func mapKCCToAPI_BucketOptions(in *krm.LogmetricBucketOptions) *api.BucketOptions {
  if (in == nil) { return nil }
  out := &api.BucketOptions{}
  out.ExplicitBuckets = mapKCCToApi_Explicit(in.ExplicitBuckets)
  out.ExponentialBuckets = mapKCCToApi_Exponential(in.ExponentialBuckets)
  out.LinearBuckets = mapKCCToApi_Linear(in.LinearBuckets)
   return out  
}

func mapKCCToAPI_Explciit(in *krm.Explicit) *api.Explicit {
  if (in == nil) { return nil }
...
```

The trick I've seen done in kubernetes (for version conversion) is to have a `<foo>_generated.go` (or something) and then only generate the functions that aren't explicitly defined.  So if you wanted to override `mapKCCToAPI_BucketOptions`, you would just have a `func mapKCCToAPI_BucketOptions(` in your non-generated code.  I think the original kube generator code used static analysis to parse the go code, but that is actually a bit of a pain because it doesn't work mid-edit when the code doesn't compile etc, so it might be more user-friendly just to regex match on e.g. `^func mapKCCToAPI_BucketOptions(` (with some escaping) and  just require people not to do things like this:

```
/* EDGE CASE */ func mapKCCToAPI_BucketOptions(
```

(although I can also see us getting cute here and removing anything in between `/*` and `*/` before doing the regexp)

cc @anhdle-sso